### ISSUE TITLE
Unexpected Group.Code value change in v4.10

### ISSUE THREAD

**Issue by JamesDPC** [silverstripe/silverstripe-framework#10238]
## Affected Version

4.10.1, 4.10.0

## Description

If a group is created in application code, then retrieved and saved with the same Group.Code value assigned, the group code is changed and gains a "-2" suffix.

While it could be seen as odd that a group code would be updated with the same value, it's also unexpected that the value is changed.

In the Group class there is a comment:

> The "Code" attribute is usually treated as a more permanent identifier than database IDs in custom application logic, so can't be changed after its first set.

https://github.com/silverstripe/silverstripe-framework/blob/4.10.1/src/Security/Group.php#L564

One result of the code value changing is that an application that relies on the code remaining unchanged will fail to retrieve the group the application has previously created.

If the group code (and titles since 4.10) are to be considered unique, then shouldn't this be enforced by a unique index in the database and picked up in validation (trying to save a group with an existing code that is not the current group's code?).

Related PR: https://github.com/silverstripe/silverstripe-framework/pull/10113

## Steps to Reproduce

```php
$code = 'test';

$group = Group::create([
    'Title' => 'Test',
    'Code' => $code,
    'Description' => 'Testing group'
]);

$group->write();

var_dump("1:" . $group->Code);

$groupSaved = Group::get()->filter( [ 'Code' => $code ] )->first();

var_dump("2:" . $groupSaved->Code);

$groupSaved->Code = $code;

var_dump("3:" . $groupSaved->Code);

$groupSaved->write();

var_dump("4:" . $groupSaved->Code);
```
Results in:

```
string(6) "1:test"
string(6) "2:test"
string(8) "3:test-2"
string(8) "4:test-2"
```

As a workaround, setCode can be bypassed  by calling setField and the Group Code remains unchanged:

```php
var_dump("2:" . $groupSaved->Code);

$groupSaved->setField('Code', $code);

var_dump("3:" . $groupSaved->Code);
```

Results in:

```
string(6) "1:test"
string(6) "2:test"
string(6) "3:test"
string(6) "4:test"
```

## ACs
- [x] When re-writing an existing group with the same code, the code is unchanged
- [x] When trying to save a new group with a code matching an existing group, a unique code is used
- [ ] Patch release 4.10.*

## PRs
- https://github.com/silverstripe/silverstripe-framework/pull/10333

**Comment by kinglozzer** [comment]
If I’m understanding correctly, the main issue here is with `setCode()` - it looks like we need `->exclude('ID', $this->ID)` (if `$this->isInDB()`). Does that sound like it’d fix this?

In my opinion we should also move the duplicate check logic from `setCode()` to `onBeforeWrite()`. No need to issue the database query before we need to 😄

**Comment by JamesDPC** [comment]
Yep it would, leaving setCode to do what it does in 4.9.x

When I enforce a unique value on write, it's something like this:

```php
$check = Thing::get()->filter([ 'Field' => $value ]);
if($this->exists()) {
    $check = $check->exclude(['ID' => $this->ID]);
}
// if $check->count() > 0 ....  validation error
```

That said, applying a unique index to the DB column is a good thing, as SQL queries writing a duplicate value will then also fail. 
(if done, that will cause build problems if duplicate values exist).

Cheers
J

**Comment by JamesDPC** [comment]
That makes sense, we can apply that via an extension in projects.

**Comment by GuySartorelli** [comment]
> In my opinion we should also move the duplicate check logic from setCode() to onBeforeWrite(). No need to issue the database query before we need to

I agree with this - however, in implementing it I've run into a problem:

1. If I move this logic into `onBeforeWrite()`, it never gets used because [the validation logic](https://github.com/silverstripe/silverstripe-framework/blob/07aae0e56af233c546f4f120e53ac6e97fea7c8e/src/Security/Group.php#L531-L539) added in #10113 kicks in and throws an exception.
2. If I leave the logic where it is, that validation logic is never run, so we get the renamed codes but there's some effectively dead code.

Because the current behaviour is to rename the code instead of throwing a validation error, I'm going to retain that behaviour and go with the first option above - suddenly throwing validation errors would be a breaking change. I do think the validation error would make more sense though. We should probably swap to that in a future major.

**Comment by JamesDPC** [comment]
Hi Guy

Thanks for this, I'll run through it.

In a future major release it would be good for the database to enforce the Group.Code as unique via an index, that's a tough one though as the index cannot be applied without existing duplicates being resolved.

Cheers
J

**Comment by GuySartorelli** [comment]
@JamesDPC This is resolved in [4.10.7](https://github.com/silverstripe/silverstripe-framework/releases/tag/4.10.7).
